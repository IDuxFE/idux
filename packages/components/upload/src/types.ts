/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { IxInnerPropTypes, IxPublicPropTypes, VKey } from '@idux/cdk/utils'
import type { ProgressProps } from '@idux/components/progress'
import type { DefineComponent, HTMLAttributes, VNode } from 'vue'

import { IxPropTypes } from '@idux/cdk/utils'

type DataType = Record<string, unknown>
export type UploadRequestHeader = Record<string, string>
export type UploadRequestMethod = 'POST' | 'PUT' | 'PATCH' | 'post' | 'put' | 'patch'
export type UploadRequestStatus = 'loadstart' | 'progress' | 'abort' | 'error' | 'loadend'
export type UploadFileStatus = 'selected' | 'uploading' | 'error' | 'success' | 'abort'
export type UploadListType = 'text' | 'image' | 'imageCard'
export type UploadIconType = 'file' | 'preview' | 'download' | 'remove' | 'retry'
export interface UploadProgressEvent extends ProgressEvent {
  percent?: number
}
export interface UploadRequestError extends Error {
  status?: number
  method?: UploadRequestMethod
  url?: string
}
export interface UploadRawFile extends File {
  uid: VKey
}
export interface UploadFile {
  uid: VKey
  name: string
  raw?: UploadRawFile
  status?: UploadFileStatus
  error?: Error
  errorTip?: string
  thumbUrl?: string
  percent?: number
  response?: Response
}
export interface UploadRequestOption<T = unknown> {
  onProgress?: (event: UploadProgressEvent) => void
  onError?: (event: UploadRequestError | ProgressEvent, body?: T) => void
  onSuccess?: (body: T) => void
  filename: string
  file: UploadRawFile | File
  withCredentials?: boolean
  action: string
  requestHeaders?: UploadRequestHeader
  requestMethod: UploadRequestMethod
  requestData?: DataType
}
export interface UploadRequestChangeOption {
  file: UploadFile
  status: UploadRequestStatus
  response?: Response
  e?: ProgressEvent
}

export const uploadProps = {
  files: IxPropTypes.array<UploadFile>().isRequired,
  accept: IxPropTypes.string,
  action: IxPropTypes.oneOfType([String, IxPropTypes.func<(file: UploadFile) => Promise<string>>()]).isRequired,
  dragable: IxPropTypes.bool,
  directory: IxPropTypes.bool,
  disabled: IxPropTypes.bool,
  maxCount: IxPropTypes.number,
  multiple: IxPropTypes.bool,
  progress: IxPropTypes.object<ProgressProps>(),
  name: IxPropTypes.string,
  customRequest: IxPropTypes.func<(option: UploadRequestOption) => { abort: () => void }>(),
  withCredentials: IxPropTypes.bool,
  requestData: IxPropTypes.oneOfType<DataType | ((file: UploadFile) => DataType | Promise<DataType>)>([
    Object,
    IxPropTypes.func<(file: UploadFile) => DataType | Promise<DataType>>(),
  ]),
  requestHeaders: IxPropTypes.object<UploadRequestHeader>(),
  requestMethod: IxPropTypes.oneOf(['POST', 'PUT', 'PATCH', 'post', 'put', 'patch']),
  'onUpdate:files': IxPropTypes.emit<(fileList: UploadFile[]) => void>(),
  onSelect: IxPropTypes.emit<(file: File[]) => boolean | File[] | Promise<boolean | File[]>>(),
  onBeforeUpload: IxPropTypes.emit<(file: UploadFile) => boolean | UploadFile | Promise<boolean | UploadFile>>(),
  onFileStatusChange: IxPropTypes.emit<(file: UploadFile) => void>(),
  onRequestChange: IxPropTypes.emit<(option: UploadRequestChangeOption) => void>(),
}
export type UploadProps = IxInnerPropTypes<typeof uploadProps>
export type UploadPublicProps = IxPublicPropTypes<typeof uploadProps>
export type UploadComponent = DefineComponent<Omit<HTMLAttributes, keyof UploadPublicProps> & UploadPublicProps>
export type UploadInstance = InstanceType<DefineComponent<UploadProps>>

export const uploadListProps = {
  type: IxPropTypes.oneOf<UploadListType>(['text', 'image', 'imageCard']),
  icon: IxPropTypes.object<Partial<Record<UploadIconType, string | VNode>>>(),
  onDownload: IxPropTypes.emit<(file: UploadFile) => void>(),
  onPreview: IxPropTypes.emit<(file: UploadFile) => void>(),
  onRemove: IxPropTypes.emit<(file: UploadFile) => boolean | Promise<boolean>>(),
  onRetry: IxPropTypes.emit<(file: UploadFile) => void>(),
}
export type UploadListProps = IxInnerPropTypes<typeof uploadListProps>
export type UploadListPublicProps = IxPublicPropTypes<typeof uploadListProps>
export type UploadListComponent = DefineComponent<
  Omit<HTMLAttributes, keyof UploadListPublicProps> & UploadListPublicProps
>
export type UploadListInstance = InstanceType<DefineComponent<UploadListProps>>
