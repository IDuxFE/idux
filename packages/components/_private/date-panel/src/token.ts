/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

/* eslint-disable @typescript-eslint/no-explicit-any */

import type { ActiveDateContext } from './composables/useActiveDate'
import type { ActiveTypeContext } from './composables/useActiveType'
import type { MaxIndexContext } from './composables/useMaxIndex'
import type { DatePanelProps } from './types'
import type { DateConfig } from '@idux/components/config'
import type { DatePickerLocale } from '@idux/components/i18n'
import type { ComputedRef, InjectionKey, Slots } from 'vue'

export interface DatePanelContext extends ActiveTypeContext, ActiveDateContext, MaxIndexContext {
  props: DatePanelProps
  slots: Slots
  locale: ComputedRef<DatePickerLocale>
  mergedPrefixCls: ComputedRef<string>
  dateConfig: DateConfig
}

export const datePanelToken: InjectionKey<DatePanelContext> = Symbol('datePanelToken')
