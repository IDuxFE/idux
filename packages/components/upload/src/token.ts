/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { UploadRequest } from './composables/useRequest'
import type { UploadFile, UploadProps } from './types'
import type { ComputedRef, InjectionKey } from 'vue'

export type UploadToken = {
  props: UploadProps
  files: ComputedRef<UploadFile[]>
  onUpdateFiles: (file: UploadFile[]) => void
  setSelectorVisible: (isShow: boolean) => void
} & UploadRequest

export const uploadToken: InjectionKey<UploadToken> = Symbol('UploadToken')
