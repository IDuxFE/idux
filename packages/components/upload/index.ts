/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { UploadComponent, UploadListComponent } from './src/types'

import UploadList from './src/List'
import Upload from './src/Upload'

const IxUpload = Upload as unknown as UploadComponent
const IxUploadList = UploadList as unknown as UploadListComponent

export { IxUpload, IxUploadList }

export type {
  UploadRawFile,
  UploadListType,
  UploadRequestMethod,
  UploadRequestStatus,
  UploadFileStatus,
  UploadProgressEvent,
  UploadFile,
  UploadRequestOption,
  UploadRequestChangeOption,
  UploadIconType,
  UploadInstance,
  UploadComponent,
  UploadPublicProps as UploadProps,
  UploadListInstance,
  UploadListComponent,
  UploadListPublicProps as UploadListProps,
} from './src/types'
