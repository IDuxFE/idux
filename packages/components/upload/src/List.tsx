/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { UploadListProps } from './types'
import type { UploadListConfig } from '@idux/components/config'

import { computed, defineComponent, h, inject, watchEffect } from 'vue'

import { useGlobalConfig } from '@idux/components/config'

import IxUploadImageCardList from './component/ImageCardList'
import IxUploadImageList from './component/ImageList'
import IxUploadTextList from './component/TextList'
import { useSelectorVisible } from './composables/useDisplay'
import { UploadToken, uploadToken } from './token'
import { uploadListProps } from './types'

const cpmMap = {
  text: IxUploadTextList,
  image: IxUploadImageList,
  imageCard: IxUploadImageCardList,
} as const

export default defineComponent({
  name: 'IxUploadList',
  props: uploadListProps,
  setup(props) {
    const config = useGlobalConfig('uploadList')
    const listType = useListType(props, config)
    const { props: uploadProps, setSelectorVisible } = inject(uploadToken, {
      props: {},
      setSelectorVisible: () => {},
    } as unknown as UploadToken)
    const [outerSelector] = useSelectorVisible(uploadProps, listType)

    watchEffect(() => setSelectorVisible(outerSelector.value))

    return () => h(cpmMap[listType.value], { ...props })
  },
})

function useListType(props: UploadListProps, config: UploadListConfig) {
  return computed(() => props.type ?? config.listType)
}
