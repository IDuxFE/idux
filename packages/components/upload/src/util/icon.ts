/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { FileOperation } from '../composables/useOperation'
import type { UploadFile, UploadIconType } from '../types'
import type { Locale } from '@idux/components/i18n'
import type { ComputedRef, VNode } from 'vue'

import { h } from 'vue'

import { isString } from 'lodash-es'

import { IxIcon } from '@idux/components/icon'

const iconMap = {
  file: 'paper-clip',
  preview: 'zoom-in',
  download: 'download',
  remove: 'delete',
  retry: 'edit',
} as const

type IconNodeType = string | boolean | VNode
type Opr = 'previewNode' | 'retryNode' | 'downloadNode' | 'removeNode'

export type IconsMap = Partial<Record<UploadIconType, Exclude<IconNodeType, true>>>

export type OprIcons = Record<Opr, VNode | null>

export function getIconNode(icon: Exclude<IconNodeType, true>): VNode | null {
  if (icon === false) {
    return null
  }
  if (isString(icon)) {
    return h(IxIcon, { name: icon })
  }
  return icon
}

export function getIcons(iconProp: Partial<Record<UploadIconType, IconNodeType>>): IconsMap {
  const iconFormat = {} as IconsMap
  let icon: UploadIconType
  for (icon in iconProp) {
    // 默认值
    if (iconProp[icon] === true) {
      iconFormat[icon] = iconMap[icon]
    } else {
      iconFormat[icon] = iconProp[icon] as Exclude<IconNodeType, true>
    }
  }
  return iconFormat
}

export function renderIcon(icon: IconsMap[keyof IconsMap] | undefined, props?: Record<string, unknown>): VNode | null {
  if (!icon) {
    return null
  }
  return h('span', props, [getIconNode(icon)])
}

export function renderOprIcon(
  file: UploadFile,
  icons: ComputedRef<IconsMap>,
  cpmClasses: ComputedRef<string>,
  fileOperation: FileOperation,
  locale: ComputedRef<Locale['upload']>,
): OprIcons {
  const previewNode = renderIcon(icons.value.preview, {
    class: `${cpmClasses.value}-icon-opr ${cpmClasses.value}-icon-preview`,
    onClick: () => fileOperation.preview(file),
    title: locale.value.preview,
  })
  const retryNode = renderIcon(icons.value.retry, {
    class: `${cpmClasses.value}-icon-opr ${cpmClasses.value}-icon-retry`,
    onClick: () => fileOperation.retry(file),
    title: locale.value.retry,
  })
  const downloadNode = renderIcon(icons.value.download, {
    class: `${cpmClasses.value}-icon-opr ${cpmClasses.value}-icon-download`,
    onClick: () => fileOperation.download(file),
    title: locale.value.download,
  })
  const removeNode = renderIcon(icons.value.remove, {
    class: `${cpmClasses.value}-icon-opr ${cpmClasses.value}-icon-remove`,
    onClick: () => fileOperation.remove(file),
    title: locale.value.remove,
  })

  return {
    previewNode,
    retryNode,
    downloadNode,
    removeNode,
  }
}
