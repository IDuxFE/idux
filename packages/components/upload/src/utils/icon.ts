/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { FileOperation } from '../composables/useOperation'
import type { UploadFile, UploadIconType } from '../types'
import type { Locale } from '@idux/components/locales'
import type { ComputedRef, VNode } from 'vue'

import { h } from 'vue'

import { isString } from 'lodash-es'

import { IxIcon } from '@idux/components/icon'

type IconNodeType = string | VNode
type Opr = 'previewNode' | 'retryNode' | 'downloadNode' | 'removeNode'

export type IconsMap = Partial<Record<UploadIconType, IconNodeType>>

export type OprIcons = Record<Opr, VNode | null>

export function getIconNode(icon: Exclude<IconNodeType, true>): VNode | null {
  if (!icon) {
    return null
  }
  if (isString(icon)) {
    return h(IxIcon, { name: icon })
  }
  return icon
}

export function renderIcon(icon: IconsMap[keyof IconsMap] | undefined, props?: Record<string, unknown>): VNode | null {
  if (!icon) {
    return null
  }
  return h('span', props, [getIconNode(icon)])
}

export function renderOprIcon(
  file: UploadFile,
  icons: ComputedRef<IconsMap>,
  cpmClasses: ComputedRef<string>,
  fileOperation: FileOperation,
  locale: Locale,
): OprIcons {
  const previewNode = renderIcon(icons.value.preview, {
    class: `${cpmClasses.value}-icon-opr ${cpmClasses.value}-icon-preview`,
    onClick: () => fileOperation.preview(file),
    title: locale.upload.preview,
  })
  const retryNode = renderIcon(icons.value.retry, {
    class: `${cpmClasses.value}-icon-opr ${cpmClasses.value}-icon-retry`,
    onClick: () => fileOperation.retry(file),
    title: locale.upload.retry,
  })
  const downloadNode = renderIcon(icons.value.download, {
    class: `${cpmClasses.value}-icon-opr ${cpmClasses.value}-icon-download`,
    onClick: () => fileOperation.download(file),
    title: locale.upload.download,
  })
  const removeNode = renderIcon(icons.value.remove, {
    class: `${cpmClasses.value}-icon-opr ${cpmClasses.value}-icon-remove`,
    onClick: () => fileOperation.remove(file),
    title: locale.upload.remove,
  })

  return {
    previewNode,
    retryNode,
    downloadNode,
    removeNode,
  }
}
