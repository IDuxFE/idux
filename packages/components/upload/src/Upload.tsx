/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { Ref } from 'vue'

import { defineComponent, provide, ref, shallowRef } from 'vue'

import { useControlledProp } from '@idux/cdk/utils'
import { IxImageViewer } from '@idux/components/image'

import FileSelector from './component/Selector'
import { useCmpClasses } from './composables/useDisplay'
import { useRequest } from './composables/useRequest'
import { uploadToken } from './token'
import { uploadProps } from './types'

export default defineComponent({
  name: 'IxUpload',
  props: uploadProps,
  setup(props, { slots }) {
    const cpmClasses = useCmpClasses()
    const [showSelector, setSelectorVisible] = useShowSelector()
    const [files, onUpdateFiles] = useControlledProp(props, 'files', [])
    const { fileUploading, abort, startUpload, upload } = useRequest(props, files)
    const { viewerVisible, images, setViewerVisible } = useImageViewer()
    provide(uploadToken, {
      props,
      files,
      fileUploading,
      onUpdateFiles,
      abort,
      startUpload,
      upload,
      setViewerVisible,
      setSelectorVisible,
    })

    return () => (
      <div class={cpmClasses.value}>
        {showSelector.value && <FileSelector>{slots.default?.()}</FileSelector>}
        {slots.list?.({ abort, upload })}
        <IxImageViewer v-model:visible={viewerVisible.value} images={images.value}></IxImageViewer>
        <div class={`${cpmClasses.value}-tip`}>{slots.tip?.()}</div>
      </div>
    )
  },
})

function useShowSelector(): [Ref<boolean>, (isShow: boolean) => void] {
  const showSelector = ref(true)

  function setSelectorVisible(isShow: boolean) {
    showSelector.value = isShow
  }

  return [showSelector, setSelectorVisible]
}

function useImageViewer() {
  const viewerVisible = ref(false)
  const images = shallowRef<string[]>([])

  function setViewerVisible(visible: boolean, imageSrc?: string): void {
    images.value = visible && imageSrc ? [imageSrc] : []
    viewerVisible.value = visible
  }

  return {
    viewerVisible,
    images,
    setViewerVisible,
  }
}
