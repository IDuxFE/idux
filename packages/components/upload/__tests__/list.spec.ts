import type { UploadFilesProps } from '../src/types'
import type { MountingOptions } from '@vue/test-utils'

import { flushPromises, mount } from '@vue/test-utils'
import { h, ref } from 'vue'

import { renderWork } from '@tests'

import { IxIcon } from '@idux/components/icon'
import { zhCN } from '@idux/components/locales'

import UploadFilesCpm from '../src/List'
import { uploadToken } from '../src/token'

const uploadFilesMount = (options?: MountingOptions<Partial<UploadFilesProps>>) => {
  const { global: { provide: { [uploadToken as unknown as string]: provideObj } = {}, ...restGlobal } = {}, ...rest } =
    options as MountingOptions<UploadFilesProps>

  return mount(UploadFilesCpm, {
    global: {
      provide: {
        [uploadToken as symbol]: {
          props: {},
          locale: zhCN,
          files: { value: [] },
          setSelectorVisible: () => {},
          ...provideObj,
        },
      },
      ...restGlobal,
    },
    ...rest,
  })
}

describe('Upload list render', () => {
  renderWork<UploadFilesProps>(UploadFilesCpm, {
    global: { provide: { [uploadToken as symbol]: { props: {}, files: { value: [] }, setSelectorVisible: () => {} } } },
  })

  test('type work', async () => {
    const wrapper = uploadFilesMount({
      global: {
        provide: {
          [uploadToken as symbol]: {
            files: ref([
              {
                key: 'test1',
                name: 'idux.svg',
                thumbUrl: '/icons/logo.svg',
              },
            ]),
          },
        },
      },
    })
    await flushPromises()

    expect(wrapper.classes()).toContain('ix-upload-list-text')

    await wrapper.setProps({ type: 'image' })

    expect(wrapper.classes()).toContain('ix-upload-list-image')

    await wrapper.setProps({ type: 'imageCard' })

    expect(wrapper.classes()).toContain('ix-upload-list-imageCard')
  })

  test('icon work', async () => {
    const wrapper = uploadFilesMount({
      global: {
        provide: {
          [uploadToken as symbol]: {
            files: ref([
              {
                key: 'test1',
                name: 'idux.svg',
                errorTip: 'error',
                status: 'error',
              },
            ]),
          },
        },
      },
    })
    await flushPromises()

    expect(wrapper.find('.ix-icon-paper-clip').exists()).toBeTruthy()
    expect(wrapper.find('.ix-icon-delete').exists()).toBeTruthy()
    expect(wrapper.find('.ix-icon-edit').exists()).toBeTruthy()

    const wrapperFileSuccess = uploadFilesMount({
      global: {
        provide: {
          [uploadToken as symbol]: {
            files: ref([
              {
                key: 'test1',
                name: 'idux.svg',
                status: 'success',
              },
            ]),
          },
        },
      },
      props: {
        icon: {
          download: 'download',
          remove: h(IxIcon, { name: 'close' }),
          file: 'left',
        },
      },
    } as MountingOptions<Partial<UploadFilesProps>>)
    await flushPromises()

    expect(wrapperFileSuccess.find('.ix-icon-download').exists()).toBeTruthy()
    expect(wrapperFileSuccess.find('.ix-icon-left').exists()).toBeTruthy()
    expect(wrapperFileSuccess.find('.ix-icon-close').exists()).toBeTruthy()
  })

  test('onDownload work', async () => {
    const onDownload = jest.fn()
    const defaultFiles = [
      {
        key: 'test1',
        name: 'idux.svg',
        status: 'success',
      },
    ]

    const wrapper = uploadFilesMount({
      global: {
        provide: {
          [uploadToken as symbol]: {
            files: ref(defaultFiles),
          },
        },
      },
      props: {
        icon: {
          download: 'download',
        },
        onDownload,
      },
    } as MountingOptions<Partial<UploadFilesProps>>)

    await wrapper.find('.ix-upload-icon-download').trigger('click')

    expect(onDownload).toBeCalledWith(expect.objectContaining(defaultFiles[0]))
  })

  test('onPreview work', async () => {
    const onPreview = jest.fn()
    const defaultFiles = [
      {
        key: 'test1',
        name: 'idux.svg',
        status: 'success',
      },
    ]

    const wrapper = uploadFilesMount({
      global: {
        provide: {
          [uploadToken as symbol]: {
            files: ref(defaultFiles),
          },
        },
      },
      props: {
        icon: {
          preview: 'zoom-in',
        },
        onPreview,
      },
    } as MountingOptions<Partial<UploadFilesProps>>)

    await wrapper.find('.ix-upload-name').trigger('click')

    expect(onPreview).toBeCalledWith(expect.objectContaining(defaultFiles[0]))

    await wrapper.setProps({ type: 'imageCard' })
    await wrapper.find('.ix-icon-zoom-in').trigger('click')

    expect(onPreview).toBeCalledWith(expect.objectContaining(defaultFiles[0]))
  })

  test('onRetry work', async () => {
    const onRetry = jest.fn()
    const upload = jest.fn()
    const defaultFiles = [
      {
        key: 'test1',
        name: 'idux.svg',
        status: 'error',
      },
    ]

    const wrapper = uploadFilesMount({
      global: {
        provide: {
          [uploadToken as symbol]: {
            files: ref(defaultFiles),
            upload,
          },
        },
      },
      props: {
        icon: {
          retry: 'edit',
        },
        onRetry,
      },
    } as MountingOptions<Partial<UploadFilesProps>>)
    await wrapper.find('.ix-upload-icon-retry').trigger('click')

    expect(upload).toBeCalled()
    expect(onRetry).toBeCalledWith(expect.objectContaining(defaultFiles[0]))
  })

  test('onRemove work', async () => {
    const onRemove = jest.fn(() => false)
    const onUpdateFiles = jest.fn()
    const abort = jest.fn()
    const defaultFiles = [
      {
        key: 'test1',
        name: 'idux.png',
        status: 'uploading',
        raw: new File(['test1'], 'idux.png', { type: 'image/png' }),
      },
    ]

    const wrapper = uploadFilesMount({
      global: {
        provide: {
          [uploadToken as symbol]: {
            files: ref(defaultFiles),
            onUpdateFiles,
            abort,
          },
        },
      },
      props: {
        onRemove,
      },
    } as MountingOptions<Partial<UploadFilesProps>>)
    await wrapper.find('.ix-upload-icon-remove').trigger('click')

    expect(onRemove).toBeCalledWith(expect.objectContaining(defaultFiles[0]))
    expect(onUpdateFiles).not.toBeCalled()
    onUpdateFiles.mockRestore()

    const allowRemove = jest.fn(() => true)
    await wrapper.setProps({ onRemove: allowRemove })
    await wrapper.find('.ix-upload-icon-remove').trigger('click')

    expect(allowRemove).toBeCalledWith(expect.objectContaining(defaultFiles[0]))
    expect(abort).toBeCalled()
    expect(onUpdateFiles).toBeCalled()
    onUpdateFiles.mockRestore()

    const allowRemovePromise = jest.fn(() => Promise.resolve(true))
    await wrapper.setProps({ onRemove: allowRemovePromise })
    await wrapper.find('.ix-upload-icon-remove').trigger('click')

    expect(allowRemovePromise).toBeCalledWith(expect.objectContaining(defaultFiles[0]))
    expect(onUpdateFiles).toBeCalled()
    expect(abort).toBeCalled()
    onUpdateFiles.mockRestore()

    const notAllowRemovePromise = jest.fn(() => Promise.resolve(false))
    await wrapper.setProps({ onRemove: notAllowRemovePromise })
    await wrapper.find('.ix-upload-icon-remove').trigger('click')

    expect(notAllowRemovePromise).toBeCalledWith(expect.objectContaining(defaultFiles[0]))
    expect(onUpdateFiles).not.toBeCalled()
  })
})
