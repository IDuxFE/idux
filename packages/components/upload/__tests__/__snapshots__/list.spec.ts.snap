// Vitest Snapshot v1

exports[`Upload list render > render work 1`] = `"<!---->"`;
